staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/samplecontroller/v1alpha1/foo.go

```diff
@@ -60,12 +71,12 @@ func NewFooInformer(client versioned.Interface, namespace string, resyncPeriod t
 	)
 }
 
-func defaultFooInformer(client versioned.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
-	return NewFooInformer(client, v1.NamespaceAll, resyncPeriod, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+func (f *fooInformer) defaultInformer(client versioned.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
```

Comment:
What if we change `NewInformerFunc` with an additional namespace arg? That would be the canonical change without the need to tweak the list watcher last minute with a filter, wouldn't it?

Reply:
I'm not sure I understand - you mean change the `defaultInformer` (or `defaultFooInformer` previously) to have a `namespace` argument?

The issue there, is that `InformerFor` in the `SharedInformerFactory` then needs to somehow be aware whether the Informer that it's constructing is for a namespaced type or not. We *could* just include `namespace` on *all* of them, and silently drop the namespace arg on non-namespaced resources.

This sort of goes against what you mentioned previously - that we could use a filter instead of just a plain string. I'd *hope* I'm not alone in wanting to be able to modify `ListOptions` of the informers I create, but perhaps I am!

Reply:
> We could just include namespace on all of them, and silently drop the namespace arg on non-namespaced resources.

That's what I had in mind. But you have a point. What is really ugly is the filter func doing two things. You could separate the namespace and the filter by:
- keep the namespace string separated in `NewFilteredFooInformer`, store it in `group` as well and pass it to `NewFilteredFooInformer` from `defaultInformer`
- rename filter to tweakListOptions.

It's not really elegant either, but less magic.

Reply:
Yeah that might be a better idea - I imagine a lot of people will want to set a namespace, but be okay with the default ListOptions. Do you think that `tweakListOptions` function should accept a pointer, or a copy of the type to then return? My only concern then was that a user should really perform a DeepCopy before modifying the copy passed into the filter, as ListOptions could at some point include some nested/pointer types. Perhaps this isn't a concern though. By passing a pointer, it's clear straight away that this type *will* be mutated.

I guess the 3rd option is to create some kind of 'Filter builder', where we can pass in a list of arbitrary functions that manipulate some kind of 'Filter' structure, that is understood by the `New*Informer` functions. This will probably be a fair amount more work, and may not be *clearer* to the user, but it'd probably be more elegant.

Reply:
(the 3rd option there obviously does not help fix the Lister inconsistency. I think we really are stuck on that one without a big change)

Reply:
> Do you think that tweakListOptions function should accept a pointer

`tweakListOptions` is explicit enough to expect side effects. So a pointer is fine.